### PR TITLE
[dev] rough linq implementation

### PR DESCRIPTION
**Pull request by justin-robinson**
*VERY* rough linq implementation.
* hardcoded tables and column names
* ignored where clause operator

built using this walkthrough https://msdn.microsoft.com/en-us/library/bb546158.aspx?cs-save-lang=1&cs-lang=csharp

---
This change is [Reviewable](https://reviewable.io/reviews/justin-robinson/crmapi/2)

### CODE REVIEW

**Comment by gooberchicken** [comment]
Looks like the tree and context and other things like that are nearly copy paste, so not sure if you wanted to review that. I've never done anything like this, so it'll take me a bit to wrap my head around it all if that's the case. 

---

Review status: 0 of 15 files reviewed at latest revision, all discussions resolved.

---

*[CRMApi/AWS/DynamoDB/Linq/Evaluator.cs, line 8 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQj3SB9FxUi5MoZHYS:-LCQj3SB9FxUi5MoZHYT:bbf1lkh) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/Evaluator.cs#L8)):*
> ```
>         public static Expression PartialEval(Expression expression, Func<Expression, bool> fnCanBeEvaluated) {
>             return new SubtreeEvaluator(new Nominator(fnCanBeEvaluated).Nominate(expression)).Eval(expression);
> ```
nitpicky, but C# curly brackets start on a new line.

---

*[CRMApi/AWS/DynamoDB/Linq/ExpressionTreeHelpers.cs, line 46 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQjkFd6m1aW-n-TO7j:-LCQjkFd6m1aW-n-TO7k:bc0irgz) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/ExpressionTreeHelpers.cs#L46)):*
> ```Smalltalk
> 
>             // We should have returned by now. 
>             throw new Exception("There is a bug in this program.");
> ```

:clap: 

---

*[CRMApi/AWS/DynamoDB/Linq/PostFinder.cs, line 12 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQjx3R7Os5SimTOaTG:-LCQjx3R7Os5SimTOaTH:bnfpia4) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/PostFinder.cs#L12)):*
> ```Smalltalk
> 
>         private Expression expression;
>         private List<string> postIds;
> ```

use guids

---

*[CRMApi/AWS/DynamoDB/Linq/PostFinder.cs, line 22 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQk2v60VQUi8zPuK3b:-LCQk2v60VQUi8zPuK3c:b-jsu484) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/PostFinder.cs#L22)):*
> ```Smalltalk
>                 if (this.postIds == null) {
>                     this.postIds = new List<string>();
>                     Visit(this.expression);
> ```

why are you visiting on a get of a list? 

---

*[CRMApi/AWS/DynamoDB/Linq/QueryContext.cs, line 25 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQl6sp96uqw3U9HvrU:-LCQl6sp96uqw3U9HvrV:b33mewf) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/QueryContext.cs#L25)):*
> ```Smalltalk
> 
>             Post[] posts;
>             if (postIds.Count == 0) {
> ```

`!postIds.Any()`

---

*[CRMApi/AWS/DynamoDB/Linq/QueryContext.cs, line 28 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQlCiF7Kc5ticxM4q9:-LCQlCiF7Kc5ticxM4qA:b-2i8wnq) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/QueryContext.cs#L28)):*
> ```Smalltalk
>                 posts = ServiceHelper.GetPosts();
>             } else {
>                 posts = ServiceHelper.GetPostsByPostId(postIds);
> ```

wait. you're calling a method called `getPostsByPostId` but sending in an array of postIds

---

*[CRMApi/AWS/DynamoDB/Linq/ServiceHelper.cs, line 27 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQmSEb3liBxI6hVibj:-LCQmSEb3liBxI6hVibk:bir1o6e) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/ServiceHelper.cs#L27)):*
> ```Smalltalk
>         private static List<Post> GetPosts(LinkedList<ScanCondition> scanConditions) {
>             List<Post> posts = new List<Post>();
>             var query = Client.GetContext().ScanAsync<Post>(scanConditions);
> ```

what is client here? does it implement `IDisposable`? 

---


*Comments from [Reviewable](https://reviewable.io/reviews/justin-robinson/crmapi/2#-:-LCQmd2v4JNlFMVFfKnR:bj1whfx)*

**Comment by justin-robinson** [comment]
Review status: 0 of 15 files reviewed at latest revision, 7 unresolved discussions.

---

*[CRMApi/AWS/DynamoDB/Linq/Evaluator.cs, line 8 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQj3SB9FxUi5MoZHYS:-LCUKWV52h4MomGGIGIQ:b-jvx6qt) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/Evaluator.cs#L8)):*
<details><summary><i>Previously, pace34cap wrote…</i></summary><blockquote>

> ```
>         public static Expression PartialEval(Expression expression, Func<Expression, bool> fnCanBeEvaluated) {
>             return new SubtreeEvaluator(new Nominator(fnCanBeEvaluated).Nominate(expression)).Eval(expression);
> ```
nitpicky, but C# curly brackets start on a new line.
</blockquote></details>

yea i changed that in the ide real quick. i didn't like the amount of wasted space.

---

*[CRMApi/AWS/DynamoDB/Linq/PostFinder.cs, line 12 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQjx3R7Os5SimTOaTG:-LCUKfl764jjuQMu7ETZ:b-7pn2eu) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/PostFinder.cs#L12)):*
<details><summary><i>Previously, pace34cap wrote…</i></summary><blockquote>

use guids
</blockquote></details>

there is a guid class?

---

*[CRMApi/AWS/DynamoDB/Linq/PostFinder.cs, line 22 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQk2v60VQUi8zPuK3b:-LCUKj437IE8EJN_roMC:b-um8jit) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/PostFinder.cs#L22)):*
<details><summary><i>Previously, pace34cap wrote…</i></summary><blockquote>

why are you visiting on a get of a list? 
</blockquote></details>

This new concept (to me) came from the walkthrough I linked. Not 100% sure what this does other than "switch/case replacement"

---

*[CRMApi/AWS/DynamoDB/Linq/QueryContext.cs, line 25 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQl6sp96uqw3U9HvrU:-LCULOGh1bhkM6XeRdDN:b-896fix) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/QueryContext.cs#L25)):*
<details><summary><i>Previously, pace34cap wrote…</i></summary><blockquote>

`!postIds.Any()`
</blockquote></details>

Done.

---

*[CRMApi/AWS/DynamoDB/Linq/QueryContext.cs, line 28 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQlCiF7Kc5ticxM4q9:-LCULPtlE3Kw3kMpR1Vk:b-xzafkk) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/QueryContext.cs#L28)):*
<details><summary><i>Previously, pace34cap wrote…</i></summary><blockquote>

wait. you're calling a method called `getPostsByPostId` but sending in an array of postIds
</blockquote></details>

yea this will change

---

*[CRMApi/AWS/DynamoDB/Linq/ServiceHelper.cs, line 27 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQmSEb3liBxI6hVibj:-LCULb3x4pl_Vytx2tJL:b-ycvxx8) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/ServiceHelper.cs#L27)):*
<details><summary><i>Previously, pace34cap wrote…</i></summary><blockquote>

what is client here? does it implement `IDisposable`? 
</blockquote></details>

It's a static shared instance of AmazonDynamoDBClient. It does have a `Dispose()` method so I'm assuming  it does implement `IDisposable`. I'm still researching where to call that method since I haven't found a reliable app destructor.
https://github.com/justin-robinson/CRMApi/blob/master/CRMApi/AWS/DynamoDB/Client.cs 

---


*Comments from [Reviewable](https://reviewable.io/reviews/justin-robinson/crmapi/2)*

**Comment by justin-robinson** [comment]
creating a queryable data source for linq. This allows me to run linq queries using DynamoDB as in input. Specifically this is a hardcoded DynamoDB Post table


---

Review status: 0 of 15 files reviewed at latest revision, 7 unresolved discussions.

---

*[CRMApi/AWS/DynamoDB/Linq/ExpressionTreeHelpers.cs, line 46 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQjkFd6m1aW-n-TO7j:-LCUMC6Y1vuNxLM7qnf4:b-myd7ji) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/ExpressionTreeHelpers.cs#L46)):*
<details><summary><i>Previously, pace34cap wrote…</i></summary><blockquote>

:clap: 
</blockquote></details>

thank the walkthrough

---


*Comments from [Reviewable](https://reviewable.io/reviews/justin-robinson/crmapi/2#-:-LCUMJK2-Ffxd4eGPEZK:bp54x69)*

**Comment by justin-robinson** [comment]
Review status: 0 of 15 files reviewed at latest revision, 7 unresolved discussions.

---

*[CRMApi/AWS/DynamoDB/Linq/PostFinder.cs, line 12 at r1](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQjx3R7Os5SimTOaTG:-LDJwKzV3RiLOtU6JEaE:b-896fix) ([raw file](https://github.com/justin-robinson/crmapi/blob/0210923cf84551cd739a2df62b1d5c4c6b190e81/CRMApi/AWS/DynamoDB/Linq/PostFinder.cs#L12)):*
<details><summary><i>Previously, justin-robinson (Justin Robinson) wrote…</i></summary><blockquote>

there is a guid class?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/justin-robinson/crmapi/2)*

**Comment by justin-robinson** [comment]
Review status: 0 of 19 files reviewed at latest revision, 7 unresolved discussions.

---

*[CRMApi/AWS/DynamoDB/Linq/ExpressionTypeHelper.cs, line 6 at r2](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LDLBUV79UMXTjF_jZLh:-LDLBUV79UMXTjF_jZLi:b84y4jt) ([raw file](https://github.com/justin-robinson/crmapi/blob/52567aee17c877a7bb6a396a40a5103ab470dc70/CRMApi/AWS/DynamoDB/Linq/ExpressionTypeHelper.cs#L6)):*
> ```Smalltalk
> 
> namespace CRMApi.AWS.DynamoDB.Linq {
>     public static class ExpressionTypeHelper {
> ```

This is how we are mapping LINQ's comparators to DynamoDb's

---

*[CRMApi/AWS/DynamoDB/Linq/ItemFinder.cs, line 7 at r2](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LDLATpnBm_ClCkNgaJo:-LDLATpnBm_ClCkNgaJp:b-dw6p8t) ([raw file](https://github.com/justin-robinson/crmapi/blob/52567aee17c877a7bb6a396a40a5103ab470dc70/CRMApi/AWS/DynamoDB/Linq/ItemFinder.cs#L7)):*
> ```Smalltalk
> 
> namespace CRMApi.AWS.DynamoDB.Linq {
>     public class ItemFinder : ExpressionVisitor {
> ```

This was the `PostFinder` class. I've genericized and renamed it

---

*[CRMApi/AWS/DynamoDB/Linq/QueryProvider.cs, line 35 at r2](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LDLA9k57kSGOI-M7oLt:-LDLA9k57kSGOI-M7oLu:b-mfzvhf) ([raw file](https://github.com/justin-robinson/crmapi/blob/52567aee17c877a7bb6a396a40a5103ab470dc70/CRMApi/AWS/DynamoDB/Linq/QueryProvider.cs#L35)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
>             var returnType = typeof(T);
>             var isEnumerable = returnType.Name == "IEnumerable`1";
>             var itemType = isEnumerable
>                 ? returnType.GenericTypeArguments[0]
>                 : returnType;
>  
>             return (T) typeof(QueryContext)
>                 .GetMethod(nameof(QueryContext.Execute), BindingFlags.Public | BindingFlags.Static)
> ```
</details>

This is what allows us to dynamically build DynamoDB object

---


*Comments from [Reviewable](https://reviewable.io/reviews/justin-robinson/crmapi/2)*

File: CRMApi/AWS/DynamoDB/Linq/InvalidQueryException.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+namespace CRMApi.AWS.DynamoDB.Linq {
+    class InvalidQueryException : Exception {
+        private string message;
+
+        public InvalidQueryException(string message) {
+            this.message = message + " ";
```
Comment (ChadMoran) [inline CRMApi/AWS/DynamoDB/Linq/InvalidQueryException.cs]: Is the extra `+ " "` needed?  I thought the base `Exception` class already had a ctor that accepted a message.

File: CRMApi/AWS/DynamoDB/Linq/ItemFinder.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        private ICollection<ScanCondition> _scanConditions;
+
+        public ItemFinder(Expression expression) {
+            _expression = expression;
+        }
+
+        public ICollection<ScanCondition> ScanConditions {
+            get {
+                if (_scanConditions != null) return _scanConditions;
+                _scanConditions = new LinkedList<ScanCondition>();
```
Comment (ChadMoran) [inline CRMApi/AWS/DynamoDB/Linq/ItemFinder.cs:19]: Generally I would put this in the ctor

File: CRMApi/Models/AbstractModel.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using CRMApi.AWS.DynamoDB.Linq;
+
+namespace CRMApi.Models {
+    public abstract class AbstractModel<T> {
+        public static QueryableServerData<T> Select () {
```
Comment (ChadMoran) [inline CRMApi/Models/AbstractModel.cs]: Style thing.  A pattern I've observed when having an abstract model is to add an `All` property so you can do `Post.All.Where(p => p)` and have other static properties that are basically prefiltered like `Post.Published.Where(p => p)`.  Also solves for yoda-LINQ.

**Comment by justin-robinson** [comment]
Review status: 0 of 20 files reviewed at latest revision, 13 unresolved discussions.

---

*[CRMApi/AWS/DynamoDB/Linq/InvalidQueryException.cs, line 7 at r3](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LCQXBy4ywAnh7CtXwkE-r3-7:-LDbrpwD80jrS-acOBKG:bwnktvg) ([raw file](https://github.com/justin-robinson/crmapi/blob/71e46907d7473f1847ec76934247cc228584318b/CRMApi/AWS/DynamoDB/Linq/InvalidQueryException.cs#L7)):*
<details><summary><i>Previously, ChadMoran (Chad Moran) wrote…</i></summary><blockquote>

Is the extra `+ " "` needed?  I thought the base `Exception` class already had a ctor that accepted a message.
</blockquote></details>

i don't think this entire constructor is necessary. this is some of the copy pasta

---

*[CRMApi/AWS/DynamoDB/Linq/ItemFinder.cs, line 19 at r3](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LDL9bcClTlwcF07OprR-r3-19:-LDbr_7b7aHFepl-ZCwQ:b-z4xep) ([raw file](https://github.com/justin-robinson/crmapi/blob/71e46907d7473f1847ec76934247cc228584318b/CRMApi/AWS/DynamoDB/Linq/ItemFinder.cs#L19)):*
<details><summary><i>Previously, ChadMoran (Chad Moran) wrote…</i></summary><blockquote>

Generally I would put this in the ctor
</blockquote></details>

that's what i would do but i thought this was a c# ism

---

*[CRMApi/Models/AbstractModel.cs, line 5 at r3](https://reviewable.io/reviews/justin-robinson/crmapi/2#-LDUfd_F_buHCRtxKx5Z-r3-5:-LDbrmyI9g1r5X_RE2Zo:b-we8ens) ([raw file](https://github.com/justin-robinson/crmapi/blob/71e46907d7473f1847ec76934247cc228584318b/CRMApi/Models/AbstractModel.cs#L5)):*
<details><summary><i>Previously, ChadMoran (Chad Moran) wrote…</i></summary><blockquote>

Style thing.  A pattern I've observed when having an abstract model is to add an `All` property so you can do `Post.All.Where(p => p)` and have other static properties that are basically prefiltered like `Post.Published.Where(p => p)`.  Also solves for yoda-LINQ.
</blockquote></details>

ohhhhh that's nice

---


*Comments from [Reviewable](https://reviewable.io/reviews/justin-robinson/crmapi/2)*

**Pull request merged**